### ISSUE TITLE
BuildStream could be more helpful when build Python code

### ISSUE THREAD

**Issue by cs-shadow** [cs-shadow/buildstream#98]
In GitLab by @bochecha on Sep 26, 2017, 07:19

I'm building Python apps with BuildStream. That is, I build:

* Python (and its dependencies) as an `autotools` element;
* a few Python modules as `pip` elements;

All of them install `*.pyc` files in `%{install-root}`.

The first 4 bytes of a `.pyc` file are a magic number identifying the version of Python for which the bytecode was generated.

The next 4 bytes are **the timestamp of the source `.py` file at the time the bytecode was generated**.

And that's where the problem is:

* I build now `foo.py` with its mtime set to today;
* the generated `foo.cpython-36.pyc` file contains the mtime of `foo.py`;
* BuildStream commits it all to OSTree;
* OSTree sets the timestamps of all files to 0;
* later, when running the `python foo.py`, Python finds that the timestamp of the `foo.py` file (0) does not match the timestamp stored inside the `foo.cpython-36.pyc` file;

This implies a pretty big penalty when starting the app, because Python will ignore the bytecode, or worse regenerate it and write it back (if the sandbox is mounted read-write).

For now I've added the following to my Python modules, and it seems to do the trick:

```yaml
config:
  post-install-commands:
  - |
    find %{install-root} -name '*.pyc' -exec python -c 'if True:
        import sys 
        filename = sys.argv[1]
        with open(filename, "rb") as f:
            bc = f.read()
        with open(filename, "wb") as f:
            f.write(bc[0:4])
            f.write(b"\0\0\0\0")
            f.write(bc[8:])
        ' '{}' \;
```

(the `if True` is only there to allow indentation, making the thing slightly more readable)

But then this seems like a general problem with Python and OSTree, so maybe it's something BuildStream itself should take care of?

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Sep 26, 2017, 08:32

Looks reasonable to me, fwiw this is also an issue for reproducible builds.

Regarding readability, i wonder if there are other tricks you could play; like defining a variable to do the python code and expanding it inside the find loop.

Finally, im sure flatpak builder has a similar thing for reproducibility purposes, lets just look and see how the approaches compare.

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Sep 26, 2017, 12:44

This morning I talked with Alex about how Flatpak deals with this problem.

The code is here: https://github.com/flatpak/flatpak-builder/blob/master/src/builder-post-process.c#L115

In the case of Flatpak, things are a bit more complex than for BuildStream, because Flatpak installs the module being built with `DESTDIR=/`. Since BuildStream installs with `DESTDIR=/buildstream/install`, some of the complexity disappears.

However, there are still a couple of corner cases which we might want to handle. The problems arise when an element installs a `.py`/`.pyc` (only one of them) file in the same path a previous element had installed both. Composing such elements together will lead to confusion at best, weird errors that are impossible to debug at worst.

Flatpak-builder has a lot of hairy code to deal with those cases (plus the ones due to how it installs everything in the same `DESTDIR`). Do we want to handle things the same way?

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Sep 26, 2017, 12:53

> The problems arise when an element installs a `.py`/`.pyc` (only one of them) file in the same path a previous element had installed both.

I realize this might not be extremely clear. So let's take a few example scenarios.

1. A installs `foo.py` and `foo.cpython-36.pyc`, B installs `foo.py` (shadowing the source from A)
2. A installs `foo.py` and `foo.cpython-36.pyc`, B installs `foo.cpython-36.pyc` (shadowing the bytecode from A)
3. A installs `foo.cpython-36.pyc` only (maybe it's a proprietary product?), B installs `foo.py` (adding a source A didn't have)
4. A installs `foo.py` only (their build system is broken and doesn't compile the bytecode?), B installs `foo.cpython-36.pyc` (maybe it compiled A's `foo.py`, maybe it's unrelated)

Quoting Alex:

> basically, its a rats nest
> and i'm not touching that code again

:stuck_out_tongue_winking_eye:

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Sep 26, 2017, 17:43

mentioned in commit 10541c428f8feb923f80566baf7ce8b13cdbdb37

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Sep 26, 2017, 17:44

mentioned in merge request !98

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Sep 29, 2017, 11:30

mentioned in commit 4a29f7eb2c53cd3d06038d3cbb44d1ce835de647

**Comment by cs-shadow** [comment]
In GitLab by @bochecha on Sep 29, 2017, 11:39

closed via commit 4a29f7eb2c53cd3d06038d3cbb44d1ce835de647

**Comment by cs-shadow** [comment]
In GitLab by @jjardon on Dec 29, 2018, 10:23

mentioned in merge request freedesktop-sdk/freedesktop-sdk!816
